### PR TITLE
Enhancement/remove cycle dependencies

### PR DESCRIPTION
**Pull request by sadovnikov**

### CODE REVIEW

**Comment by sadovnikov** [comment]
The major updates are:
* `MesosCluster` and entire `com.containersol.minimesos.cluster` package depend on interfaces for members of the cluster instead of their concrete implementations;
* `MesosClusterFactory` interface define how to deserialise and to destroy the running cluster;
* CLI commands and unit tests use `MesosClusterContainersFactory` implementation of `MesosClusterFactory`, which uses docker
* `MesosCluster` is separated from JUnit `TestRule`, therefore unit tests start using `MesosClusterTestRule` extension of `MesosCluster`

File: minimesos/src/main/java/com/containersol/minimesos/cluster/ClusterMember.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    void start(int timeout);
+
+    String getRole();
+
+    /**
+     * Removes a container with force
+     */
+    void remove();
+
+}
```
Comment (frankscholten) [inline minimesos/src/main/java/com/containersol/minimesos/cluster/ClusterMember.java]: Eventually I think we should refactor this to `ClusterProcess` and change container references to process references so it is clear this is a process that can be started, stopped and removed. This also shows that it can have different implementations, either container based or not. We can do it here or in a separate PR.

Comment (sadovnikov) [inline minimesos/src/main/java/com/containersol/minimesos/cluster/ClusterMember.java]: @frankscholten this renaming would require changing a lot of files again. Therefore it's done now

File: minimesos/src/main/java/com/containersol/minimesos/cluster/Filter.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        return abstractContainer -> abstractContainer instanceof MesosMaster;
+    }
+
+    public static Predicate<ClusterProcess> mesosAgent() {
+        return abstractContainer -> abstractContainer instanceof MesosAgent;
+    }
+
+    public static Predicate<ClusterProcess> marathon() {
+        return abstractContainer -> abstractContainer instanceof Marathon;
+    }
```
Comment (frankscholten) [inline minimesos/src/main/java/com/containersol/minimesos/cluster/Filter.java]: Refactor `abstractContainer` identifier to `process`

File: minimesos/src/main/java/com/containersol/minimesos/cluster/MesosCluster.java [hunk 3]
```diff
@@ -61,25 +37,20 @@
 
     private final ClusterConfig clusterConfig;
 
-    private List<AbstractContainer> containers = Collections.synchronizedList(new ArrayList<>());
+    private List<ClusterProcess> members = Collections.synchronizedList(new ArrayList<>());
```
Comment (frankscholten) [inline minimesos/src/main/java/com/containersol/minimesos/cluster/MesosCluster.java]: Refactor to `clusterProcesses`

File: minimesos/src/main/java/com/containersol/minimesos/cluster/MesosCluster.java [hunk 4]
```diff
@@ -61,25 +37,20 @@
...
-        this.containers = clusterArchitecture.getClusterContainers().getContainers();
-        this.clusterConfig = clusterArchitecture.getClusterConfig();
+    public MesosCluster(ClusterConfig clusterConfig, List<ClusterProcess> members) {
+        this.members = members;
+        this.clusterConfig = clusterConfig;
 
         clusterId = Integer.toUnsignedString(new SecureRandom().nextInt());
-        for (AbstractContainer container : containers) {
+        for (ClusterProcess container : members) {
             container.setCluster(this);
```
Comment (frankscholten) [inline minimesos/src/main/java/com/containersol/minimesos/cluster/MesosCluster.java]: Refactor to `process`

File: minimesos/src/main/java/com/containersol/minimesos/cluster/MesosCluster.java [hunk 5]
```diff
@@ -265,9 +164,9 @@ public void state(PrintStream out, String agentContainerId) {
     public void stop() {
         LOGGER.debug("Cluster " + getClusterId() + " - stop");
 
-        if (containers.size() > 0) {
-            for (int i = containers.size() - 1; i >= 0; i--) {
-                AbstractContainer container = containers.get(i);
+        if (members.size() > 0) {
+            for (int i = members.size() - 1; i >= 0; i--) {
+                ClusterProcess container = members.get(i);
                 LOGGER.debug("Removing container [" + container.getContainerId() + "]");
```
Comment (frankscholten) [inline minimesos/src/main/java/com/containersol/minimesos/cluster/MesosCluster.java]: Refactor to process

File: minimesos/src/main/java/com/containersol/minimesos/cluster/MesosCluster.java [hunk 6]
```diff
@@ -403,57 +297,31 @@ public JSONObject getAgentStateInfo(String containerId) {
...
-                DockerClientFactory.build().removeContainerCmd(container.getId()).withForce().withRemoveVolumes(true).exec();
-            }
-        }
-        LOGGER.info("Destroyed minimesos cluster " + clusterId);
-    }
-
-    public List<AbstractContainer> getContainers() {
-        return containers;
+    public List<ClusterProcess> getMembers() {
+        return members;
```
Comment (frankscholten) [inline minimesos/src/main/java/com/containersol/minimesos/cluster/MesosCluster.java]: Refactor to `getProcesses()`

File: minimesos/src/main/java/com/containersol/minimesos/cluster/ClusterMember.java (hunk 1 above)
Comment (frankscholten) [inline minimesos/src/main/java/com/containersol/minimesos/cluster/ClusterMember.java]: Ok cool. Have a look at some of the identifiers. They are still use `container` instead of `process`

**Comment by frankscholten** [comment]
LGTM

**Comment by alex-glv** [comment]
My tests fail.
```:minimesos:test

com.containersol.minimesos.ConsulRegistrationTest > testRegisterServiceWithConsul FAILED
    java.lang.AssertionError at ConsulRegistrationTest.java:50
```

Anyone else?

File: minimesos/src/main/java/com/containersol/minimesos/main/factory/MesosClusterContainersFactory.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        List<Container> dockerContainers = DockerClientFactory.build().listContainersCmd().exec();
+        Collections.sort(dockerContainers, (c1, c2) -> Long.compare(c1.getCreated(), c2.getCreated()));
+
+        for (Container container : dockerContainers) {
+            String name = ContainerName.getFromDockerNames(container.getNames());
+            if (ContainerName.belongsToCluster(name, clusterId)) {
+
+                String containerId = container.getId();
+                String[] parts = name.split("-");
+                String role = parts[1];
```
Comment (alex-glv) [inline minimesos/src/main/java/com/containersol/minimesos/main/factory/MesosClusterContainersFactory.java]: This could benefit from checking for IndexOutOfBoundException.

**Comment by sadovnikov** [comment]
debugging of the failing test... although it was not failing

image

```
vikmac:mini-mesos vik$ docker ps -a
CONTAINER ID        IMAGE                                                COMMAND                  CREATED             STATUS              PORTS                                                                                 NAMES
c54d01c6caff        tutum/hello-world                                    "/bin/sh -c 'php-fpm "   2 minutes ago       Up 2 minutes        80/tcp                                                                                helloworld-4038784688
825602bae4d9        gliderlabs/registrator:v6                            "/bin/registrator -in"   2 minutes ago       Up 2 minutes                                                                                              minimesos-registrator-1681934897-1235884795
01e4682478c1        containersol/consul-server:0.6                       "/bin/consul agent -s"   3 minutes ago       Up 3 minutes        8300-8302/tcp, 8400/tcp, 8500/tcp, 8600/tcp, 8301-8302/udp, 172.17.0.1:53->8600/udp   minimesos-consul-1681934897-1397323526
87fea311c5a6        mesosphere/marathon:v0.15.3                          "./bin/start --master"   3 minutes ago       Up 3 minutes        8080/tcp                                                                              minimesos-marathon-1681934897-1061444681
53358ecee3eb        containersol/mesos-agent:0.25.0-0.2.70.ubuntu1404    "/usr/local/bin/mesos"   3 minutes ago       Up 3 minutes        5051/tcp, 31000-32000/tcp                                                             minimesos-agent-1681934897-2203397533
21226217b2f4        containersol/mesos-master:0.25.0-0.2.70.ubuntu1404   "mesos-master --regis"   3 minutes ago       Up 3 minutes        5050/tcp                                                                              minimesos-master-1681934897-2676949481
58e3ab8c64c3        jplock/zookeeper:3.4.6                               "/opt/zookeeper/bin/z"   3 minutes ago       Up 3 minutes        2181/tcp, 2888/tcp, 3888/tcp                                                          minimesos-zookeeper-1681934897-2958213842
```

**Comment by alex-glv** [comment]
LGTM
